### ISSUE TITLE
[Feat]: Add Nvidia GPU Voltage to metrics

### ISSUE THREAD

**Issue by 8465231** [netdata/netdata#14202]
### Description

It would be VERY handy for certain workloads like GPU mining to be able to see the GPU voltage.

This is easily seen with nvidiai-smi using this command:

nvidia-smi -q -d voltage

Is it possible to add this to the nvidia-smi plugin?

### Importance

nice to have

### Value proposition

1. Allows you to dial in the optimal efficiency, particularly with multiple cards.
2. Diagnose issues with partial crashes causing odd issues that can reset the voltage setting
3. Diagnose what causes crashes since the voltage is often related to crashes.
4. The plugin for nvidia-smi already exists, it should not be that hard to add one more metric I would think?
...

### Proposed implementation

Since it can already query nvidia-smi I think all that would be needed is adding a new graph and pooling this command to populate it: 

nvidia-smi -q -d voltage

Thanks for the consideration.

**Comment by ilyam8** [comment]
Hi, @8465231.

> nvidia-smi -q -d voltage

- I see there is voltage in the [nvidia-smi -x -q output](https://github.com/netdata/go.d.plugin/blob/7cc5a789a5e04977e9d274e7c28653e6b4070bca/modules/nvidia_smi/testdata/tesla-p100.xml#L236-L238). 
- Unfortunately, this metric is not available when using [CSV format](https://github.com/netdata/go.d.plugin/tree/master/modules/nvidia_smi#configuration) which is the default (fewer metrics, but consumes much less CPU time).

Ok, I will add voltage, but I am going to add it to the [go.d/nvidia_smi](https://github.com/netdata/go.d.plugin/tree/master/modules/nvidia_smi#nvidia-gpu-monitoring-with-netdata) (python version will be deprecated in the nearest future).

**Comment by ilyam8** [comment]
Will be available in the nightly release on Saturday or early next week.

**Comment by 8465231** [comment]
I can't figure out how to change to the XML nvidia-smi version.

It says
`"Edit the go.d/nvidia_smi.conf configuration file using edit-config from the Netdata [config directory](https://learn.netdata.cloud/docs/configure/nodes), which is typically at /etc/netdata"`

But it just gives an error:

```
/etc/netdata# ./edit-config go.d/nvidia-smi.conf
File 'go.d/nvidia-smi.conf' is not found in '/usr/lib/netdata/conf.d'
```

If I just do this command: ./edit-config go.d.conf

The config that comes up does not have the lined mentioned in github leading me to believe it is the wrong file unless I am supposed to add them there?

**Comment by ilyam8** [comment]
> /etc/netdata# ./edit-config go.d/nvidia-smi.conf

It is `nvidia_smi.conf` (with `_`, not `-`).

**Comment by ilyam8** [comment]
It also needs to be enabled in `go.d.conf`.

**Comment by ilyam8** [comment]
I 🤷‍♂️ why it is "not found". We [have it](https://github.com/netdata/go.d.plugin/blob/master/config/go.d/nvidia_smi.conf) and it is installed along with other go.d config files.

Clean install from source:

```bash
pve-deb-work ~ $ ls -l /opt/netdata/usr/lib/netdata/conf.d/go.d | wc -l
79
pve-deb-work ~ $ ls -l /opt/netdata/usr/lib/netdata/conf.d/go.d | grep nvidia
.rw-r--r-- 2.0k root 31 Dec  2022 nvidia_smi.conf
```

**Comment by ilyam8** [comment]
What is your Netdata version? Install method?

**Comment by 8465231** [comment]
> I 🤷‍♂️ why it is "not found". We [have it](https://github.com/netdata/go.d.plugin/blob/master/config/go.d/nvidia_smi.conf) and it is installed along with other go.d config files.
> 
> Clean install from source:
> 
> ```shell
> pve-deb-work ~ $ ls -l /opt/netdata/usr/lib/netdata/conf.d/go.d | wc -l
> 79
> pve-deb-work ~ $ ls -l /opt/netdata/usr/lib/netdata/conf.d/go.d | grep nvidia
> .rw-r--r-- 2.0k root 31 Dec  2022 nvidia_smi.conf
> ```

Hmm, no idea how to install from source.

I just followed the install guide here: https://learn.netdata.cloud/docs/agent/packaging/installer

Everything seems to work great besides these missing files. 

It is version 1.37.1

**Comment by ilyam8** [comment]
Create `nvidia_smi.conf` manually

```yaml
jobs:
  - name: nvidia_smi
    use_csv_format: no
```

**Comment by 8465231** [comment]
> Create `nvidia_smi.conf` manually
> 
> ```yaml
> jobs:
>   - name: nvidia_smi
>     use_csv_format: no
> ```

Ok, done, I guess just wait until next week when the update is pushed and see if the voltage shows up to know if it is working?

Thanks for the help!

**Comment by ilyam8** [comment]
>  until next week

Should be tomorrow.

**Comment by 8465231** [comment]
Great! Thanks again!

I have been wanting a way to monitor GPU voltage like windows for a long time!

**Comment by 8465231** [comment]
> > It is version 1.37.1
> 
> Correction - you are using the stable version, not nighty. So no updates till the next stable release. You need:
> 
>     * build go.d.plugin manually.
> 
>     * replace the current version with the built version.
> 
> 
> The easiest way is to use docker:
> 
> ```shell
> # clone repo
> git clone https://github.com/netdata/go.d.plugin
> 
> # build go.d.plugin in a docker container (so you don't need to install golang on your system)
> docker  run -i --rm -v $PWD/go.d.plugin:/go/plugin golang:1.19-bullseye sh -c 'cd plugin && hack/go-build.sh'
> 
> # stop netdata, we need it to replace go.d.plugin
> sudo systemctl stop netdata
> 
> # replace go.d.plugin
> sudo cp go.d.plugin/bin/godplugin /usr/libexec/netdata/plugins.d/go.d.plugin
> 
> # start netdata
> sudo systemctl start netdata
> 
> # cleanup
> rm -rf go.d.plugin/
> docker rmi golang:1.19-bullseye
> ```

I reinstalled with the script and selected nightly, is that a valid method for switching to nightly?

If not I can do that manually.

**Comment by ilyam8** [comment]
I think so. Check the version - should be ~ v1.37.0-103.

**Comment by 8465231** [comment]
Says it is:

netdata v1.37.1

I am confused, it seems newer then the nightly?

**Comment by ilyam8** [comment]
tl;dr - no, in fact, it is not. Long answer - https://github.com/netdata/netdata/issues/14112.

**Comment by 8465231** [comment]
Ok, I managed to completely bork everything somehow. Tried to run the uninstall script but it ended with an error and now I can neither uninstall, reinstall or install anything anymore.

```
 WARNING  Found an existing netdata install managed by the system package manager, but could not determine the install type. Usually this means you installed an unsupported third-party netdata package.

Attempting to update an installation managed by the system package manager is known to not work in most cases. If you are trying to install the latest version of Netdata, you will need to manually uninstall it through your system package manager. If you just want to claim this install, you should re-run this command with the --claim-only option instead. Are you sure you want to continue? [y/n]
y
 --- OK, continuing ---
 WARNING  Could not find a usable copy of the updater script. We are unable to update this system in place.

 WARNING  Failed to update existing Netdata install at /.

 --- Not attempting to claim existing install at / (no claiming token provided). ---
The following non-fatal warnings or errors were encountered:

  - Found an existing netdata install managed by the system package manager, but could not determine the install type. Usually this means you installed an unsupported third-party netdata package.
  - Could not find a usable copy of the updater script. We are unable to update this system in place.
  - Failed to update existing Netdata install at /.
```
It keeps detecting a netdata install but will not uninstall it nor install over it.

Is my only option at this point to completely wipe the system and start over (not idea since this system ideally should not be taken offline)?

**Comment by ilyam8** [comment]
What is your distro?

**Comment by 8465231** [comment]
ubuntu 20.04

**Comment by ilyam8** [comment]
Try this:

```
sudo systemctl stop netdata
sudo apt-get purge netdata
sudo rm -rf /var/log/netdata/
sudo rm -rf /var/cache/netdata/
sudo rm -rf /var/lib/netdata/
sudo rm -rf /etc/netdata/
sudo userdel -r netdata
```

**Comment by 8465231** [comment]
Thanks got it working on nightly now! Just got to wait for the release to drop lol.

Thanks again!

**Comment by 8465231** [comment]
> > until next week
> 
> Should be tomorrow.

I am guessing that it did not get pushed in yesterdays update?

Says I have version netdata v1.37.0-107-nightly

I do not see any new metrics under nvidia-smi though?

Nvidia-smi is enabled in python.d.conf
Nvidia-smi is enabled in go.d.conf
and /etc/netdata/edit-config nvidia_smi.conf was added with the following commands:

jobs:
  - name: nvidia_smi
    use_csv_format: no

**Comment by ilyam8** [comment]
> Nvidia-smi is enabled in python.d.conf

Try to disable it. Only one collector with this name can be active at a time. So, if python nvidia_smi works, then go nvidia_smi doesn't.

**Comment by 8465231** [comment]
> > Nvidia-smi is enabled in python.d.conf
> 
> Try to disable it. Only one collector with this name can be active at a time. So, if python nvidia_smi works, then go nvidia_smi doesn't.

REALLY sorry to be such a bother but I disabled the python.d.conf nvidia-smi and restarted netdata, now I do not see nvidia-smi at all in netdata?

I am on version: netdata v1.37.0-115-nightly

**Comment by ilyam8** [comment]
@8465231 usually checking logs is a good first step to debug a problem

> grep nvidia_smi /var/log/netdata/error.log

**Comment by 8465231** [comment]
> @8465231 usually checking logs is a good first step to debug a problem
> 
> > grep nvidia_smi /var/log/netdata/error.log

You would think so right? That returns nothing lol.

I checked out that log fine and it is not very long and just has basic DBengine info and a few warnings about fireqos and a helper.conf.

Double checked and `/etc/netdata/edit-config nvidia_smi.conf`

has

```
jobs:
  - name: nvidia_smi
    use_csv_format: no
```

and this file has nvidia-smi : yes 
`/etc/netdata/edit-config go.d.conf`

Nividia-smi is commented out in the python.d file

**Comment by ilyam8** [comment]
> and this file has nvidia-smi : yes

Did you uncomment it (remove `#` before nvidia_smi)?

> You would think so right? That returns nothing lol.

Probably logs got rotated. Restart netdata and try again.

Additionally, see https://github.com/netdata/go.d.plugin/tree/master/modules/nvidia_smi#troubleshooting

**Comment by 8465231** [comment]
> Did you uncomment it (remove `#` before nvidia_smi)?

Yep, of course ;-)

> Probably logs got rotated. Restart netdata and try again.

Spot on as usual.

```
2023-01-11 06:14:49: go.d INFO: main[main] looking for 'nvidia_smi.conf' in [/etc/netdata/go.d /usr/lib/netdata/conf.d/go.d]
2023-01-11 06:14:49: go.d INFO: main[main] found '/etc/netdata/go.d/nvidia_smi.conf
2023-01-11 06:14:49: go.d ERROR: prometheus[nvidia_smi_exporter_local] Get "http://127.0.0.1:9454/metrics": dial tcp 127.0.0.1:9454: connect: connection refused
2023-01-11 06:14:49: go.d ERROR: prometheus[nvidia_smi_exporter_local] check failed
2023-01-11 06:14:50: python.d INFO: plugin[main] : [nvidia_smi] is disabled by default, skipping it
```

I have no idea what would cause that error?

Here is the debug output, nothing stands out to me?

```
[ DEBUG ] main[main] main.go:113 plugin: name=go.d, version=v0.49.0
[ DEBUG ] main[main] main.go:115 current user: name=netdata, uid=130
[ INFO  ] main[main] agent.go:136 instance is started
[ INFO  ] main[main] setup.go:42 loading config file
[ INFO  ] main[main] setup.go:50 looking for 'go.d.conf' in [/etc/netdata /usr/lib/netdata/conf.d]
[ INFO  ] main[main] setup.go:57 found '/etc/netdata/go.d.conf
[ INFO  ] main[main] setup.go:64 config successfully loaded
[ INFO  ] main[main] agent.go:140 using config: enabled 'true', default_run 'true', max_procs '0'
[ INFO  ] main[main] setup.go:69 loading modules
[ INFO  ] main[main] setup.go:88 enabled/registered modules: 1/79
[ INFO  ] main[main] setup.go:93 building discovery config
[ INFO  ] main[main] setup.go:123 looking for 'nvidia_smi.conf' in [/etc/netdata/go.d /usr/lib/netdata/conf.d/go.d]
[ INFO  ] main[main] setup.go:139 found '/etc/netdata/go.d/nvidia_smi.conf
[ INFO  ] main[main] setup.go:144 dummy/read/watch paths: 0/1/0
[ INFO  ] discovery[manager] manager.go:92 registered discoverers: [file discovery: [file reader]]
[ INFO  ] discovery[manager] manager.go:97 instance is started
[ INFO  ] run[manager] run.go:32 instance is started
[ INFO  ] discovery[file manager] discovery.go:73 instance is started
[ INFO  ] build[manager] build.go:107 instance is started
[ INFO  ] discovery[file reader] read.go:41 instance is started
[ INFO  ] discovery[file reader] read.go:42 instance is stopped
[ DEBUG ] build[manager] build.go:154 received config group ('/etc/netdata/go.d/nvidia_smi.conf'): 0 jobs (added: 0, removed: 0)
[ DEBUG ] run[manager] run.go:43 tick 0
[ DEBUG ] run[manager] run.go:43 tick 1
[ DEBUG ] run[manager] run.go:43 tick 2
[ DEBUG ] run[manager] run.go:43 tick 3
[ DEBUG ] run[manager] run.go:43 tick 4
[ DEBUG ] run[manager] run.go:43 tick 5
```

**Comment by ilyam8** [comment]
> [ DEBUG ] build[manager] build.go:154 received config group ('/etc/netdata/go.d/nvidia_smi.conf'): 0 jobs (added: 0, removed: 0)

Can you `cat /etc/netdata/go.d/nvidia_smi.conf`?

**Comment by 8465231** [comment]
> > [ DEBUG ] build[manager] build.go:154 received config group ('/etc/netdata/go.d/nvidia_smi.conf'): 0 jobs (added: 0, removed: 0)
> 
> Can you `cat /etc/netdata/go.d/nvidia_smi.conf`?

Hmm, I tried to use the edit-config program to edit that file since that appears to be the correct way to do things but seems that it put it in the wrong place, it was just in /etc/netdata instead of the /go.d folder.

I moved it to the go.d folder and sure enough everything started working! Got the voltages and already can spot room for improvement!

Thanks again and sorry for all the hassle!

**Comment by ilyam8** [comment]
> and already can spot room for improvement

share?

**Comment by 8465231** [comment]
> > and already can spot room for improvement
> 
> share?

lol, I was talking about improvement for my overclocks to make them more efficient. a few cards are running noticeably higher voltages then the others so I should be able to drop them down some more.

Also another card that has been more crash prone is running lower voltage then the other for some reason, like below the minimum voltage that is supposed to be possible, so gonna bump it up some.

Not a major thing but used properly can help in these kind of situations where trying to maximize efficiency. When mining was profitable we just jacked the power for more hashrate and didn't really worry about efficiency, not so anymore lol.

**Comment by ilyam8** [comment]
Ah, I was under the impression you mean to improve the collector 🙈.
